### ISSUE TITLE
In refine instance mode, starting a new instance forgets the one in progress

### ISSUE THREAD

**Issue by coqbot** [coq/coq#3890]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#3890
From: @mikeshulman
Reported version: 8.5
CC:   @gares, @JasonGross, @maximedenes

See also: [BZ#4638](https://github.com/coq/coq/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%234638%22)

**Comment by ejgallego** [comment]
@maximedenes can you see if this solved, and if not what remains to be done?

**Comment by maximedenes** [comment]
Fixed by https://github.com/coq/coq/pull/9530.

**Comment by maximedenes** [comment]
In fact no, there is still an STM bug:

```
Class Foo.
Class Bar := b : Type.

Set Nested Proofs Allowed.
Instance foo : Foo.
(* 1 subgoals, subgoal 1 (ID 4)
  
  ============================
   Foo *)

Instance bar : Bar.
exact Type.
Defined.
(* bar is defined *)

About foo.
(* foo not a defined object. *)

Defined.
(** Error: No focused proof (No proof-editing in progress). *)

Backtrack 12 0 0.
(*
Error: Anomaly "edit_at 12: Uncaught exception Stm.Vcs_aux.Expired.."
Please report at http://coq.inria.fr/bugs/.
*)
```

Cc @gares

**Comment by SkySkimmer** [comment]
Cannot reproduce

**Comment by maximedenes** [comment]
Really? I get it consistently with coqtop, master at 5c5bd952e9c28c3acf740fcdced03b2b7145076d.

**Comment by ppedrot** [comment]
Can still reproduce on a recent master 302334cd0e3961a8476e8071ca4482318e0dce6f. But the validity the new avatar of bug is arguable, in my opinion. `BackTo` is a coqtop-only command that cannot be used from UIs and is basically a pointy stick to poke at dangerous places. I'm inclined to close this issue since the original behaviour has been fixed.

**Comment by ejgallego** [comment]
The bug is valid and it impacts all IDEs. In Maxime example, the proof state is threaded incorrectly and lost, resulting in an invalid configuration.

**Comment by gares** [comment]
This bug is about the stm. If it has no more anything to do with BackTo it should ve affected to another component.

**Comment by ejgallego** [comment]
Indeed, tags adjusted. The bug is reproduced without the STM with the original script:

```coq
Class Foo.
Class Bar := b : Type.

Instance foo : Foo.
(* 1 subgoals, subgoal 1 (ID 4)
  
  ============================
   Foo *)

Instance bar : Bar.
exact Type.
Defined.
(* bar is defined *)

About foo.
(* foo not a defined object. *)

Defined.
(** Error: No focused proof (No proof-editing in progress). *)
```
So indeed here is the bug that is described in the title (and still open)

Then the backto anomaly could indeed be an stm bug, but as the vernac layer messes with the proof state wrongly, I'm unsure if the STM problem would still be there once the other bug is fixed.

The bug was introduced in 8.4 -> 8.5 but likely as part of the proof layer changes (as a few others similar to this one, i.e. #9638 )

**Comment by SkySkimmer** [comment]
Is there really still a bug? I can't reproduce any issue here. Please give the full commands used to get a bug.

**Comment by SkySkimmer** [comment]
The script works fine for me (with Set Nested Proofs Allowed added, otherwise Instance bar fails with the nested proofs error).

**Comment by ejgallego** [comment]
It actually works for me on `master` now too, maybe some of the recent work fixed some classification problem?

**Comment by SkySkimmer** [comment]
Fixed by #16278
